linbox/algorithms/polynomial-matrix/order-basis.h

```diff
@@ -347,7 +348,7 @@ namespace LinBox {
                 size_t M_Basis(PMatrix1              &sigma,
                                const PMatrix2        &serie,
                                size_t                 order,
-                               std::vector<size_t>   &shift)
+                               std::vector<size_t>   &shift2) // shift should be a BlasVector -> done by copying shift2 to shift
```

Comment:
Can't we declare BlasVector in the callers?

Reply:
I think it opens up a debate. Do we want in LinBox to manage any vectors through BlasVector ?

My opinion is that it is not a good idea. It is quite annoying to create a ZRing<uin64_t> in order to create a vector of int64_t. Furthermore, it would imply that any operations on vector element must be done through the given ring. In that specific case, shifts are just updated with +1 operation. The only BlasVector needs is to apply a BlasPermutation to the vector.

My proposition for BlasMatrixDomain is quite strict and it does not  consider std::vector as correct input vector (no sense as it is not attached to a field). Therefore, when one wants to multiply a std::vector with a BlasPermutation it as to convert the vector to BlasVector. Another solution would be to allow std::vector as input to BlasMatrixDomain operations but it will cause a lot of duplicate codes and also it breaks the Field encapsulation rules.
An intermediate solution would be to allow only a multiplication between a std::vector and a BlasPermutation even if it break the API rules it cannot cause any trouble as no field operations are needed.

Wait for any other opinions.

Reply:
OK. I like the intermediate solution.

Also BlasPermutation needs a field. with commit 5da5762f7 I propose to make it static but I think I do not like that. Any thoughts?

Reply:
Unfortunately, the intermediate solution is not feasible.
Indeed, FFLAS-FFPACK applyP code use a field argument that is attached to the elements that are permuted. Therefore, there is not way to retrieve the field from a std::vector container. Of course it is different from the one in BlasPermutation.